### PR TITLE
Change to.dtype_layout to c10-full

### PR DESCRIPTION
**Pull request by smessmer**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#41169 Change to.dtype_layout to c10-full**

-

Differential Revision: [D22289257](https://our.internmc.facebook.com/intern/diff/D22289257/)

### CODE REVIEW

File: aten/src/ATen/test/basic.cpp [hunk 1]
```diff
@@ -361,24 +361,6 @@ TEST(BasicTest, FactoryMethodsTest) {
...
-
-  // Test setting requires_grad to true.
-  tensor0 = at::empty({4}, at::TensorOptions().requires_grad(true));
-  ASSERT_EQ(tensor0.dtype(), at::kFloat);
-  ASSERT_EQ(tensor0.layout(), at::kStrided);
-  ASSERT_EQ(tensor0.device(), at::kCPU);
-  // This is a bug. Requires_grad was set to TRUE but this is being ignored.
-  // Issue https://github.com/pytorch/pytorch/issues/30405
-  ASSERT_FALSE(tensor0.requires_grad());
-  ASSERT_FALSE(tensor0.is_pinned());
```
Comment (ezyang) [inline aten/src/ATen/test/basic.cpp:380]: My recommendation is we fix #30405 first and then revisit this patch.

File: aten/src/ATen/core/op_registration/hacky_wrapper_for_legacy_signatures.h [hunk 2]
```diff
@@ -15,6 +15,9 @@ namespace c10 {
 namespace impl {
 
 inline c10::optional<MemoryFormat> process_memory_format(const TensorOptions& options, c10::optional<MemoryFormat> memory_format) {
+    TORCH_CHECK(options.requires_grad_opt() == c10::nullopt,
+        "Operators taking TensorOptions cannot take a TensorOptions with "
+        "options.requires_grad set as ", options.requires_grad());
```
Comment (ezyang) [inline aten/src/ATen/core/op_registration/hacky_wrapper_for_legacy_signatures.h:20]: It's weird to do a requires grad opt test in `process_memory_format`

Comment (smessmer) [reply aten/src/ATen/core/op_registration/hacky_wrapper_for_legacy_signatures.h:20]: True. It's unfortunately the only place where I can add it without big codegen changes. Open to ideas.

Comment (ezyang) [reply aten/src/ATen/core/op_registration/hacky_wrapper_for_legacy_signatures.h:20]: Rename `process_memory_format` then?

File: aten/src/ATen/test/basic.cpp (hunk 1 above)
Comment (smessmer) [reply aten/src/ATen/test/basic.cpp:380]: I changed it to keep the requires_grad(false) test as is and changed the requires_grad(true) test to an EXPECT_ANY_THROW.

**Review by ezyang — APPROVED** [review]

**Pull request closed without merging**